server/src/main/java/com/microsoft/java/bs/core/internal/gradle/GradleApiConnector.java

```diff
@@ -65,7 +68,13 @@ public String getGradleVersion(URI projectUri) {
    * @param client connection to BSP client
    * @return an instance of {@link GradleSourceSets}
    */
-  public GradleSourceSets getGradleSourceSets(URI projectUri, BuildClient client) {
+  public GradleSourceSets getGradleSourceSets(
+      URI projectUri,
+      BuildClient client,
+      final CancelChecker cancelChecker,
+      final CancellationToken token) throws CancellationException {
+    cancelChecker.checkCanceled();
```

Comment:
It looks not the right way to do the cancel check.

My understanding about this change is:

We checked the cancel status at the entry of the method, and may have two result

- cancelChecker.checkCanceled() throws CancellationException. In such case, the below code will not be executed, which is fine.
- Nothing happens because of the reuqest is not cancelled. Then code will continue to be executed. But, please be aware that, in this case, no one will call `tokenSource.cancel();`. If the request is cancelled after this point, the gradle will still run its tasks without cancellation. In other words, the `.withCancellationToken(token)` does not take any effect.

Reply:
Actually we're checking cancellation in first line for all the stages and we're using the same CancellationTokenSource for all Gradle Operations. So if cancellation was requested after checking cancelChecker.checkCanceled() in the current stage it'll be caught in the subsequent stage and when we call tokenSource.cancel() all the Gradle Operations involved in the prior stage will be cancelled.

This can only be an issue when this happens in the last stage. So to guard this, maybe can check for cancellation in the start and end as well for all stages.

Reply:
Sorry I'm little bit confused here. Could you help me understand more about this example?

Let's say the code passes line 76. And execute to line 103: `customModelBuilder.withCancellationToken(token).get().`. In such case, how will the customModelBuilder get be cancelled?

Reply:
public void reloadWorkspace(final CancelChecker cancelChecker,
      final CancellationToken token) throws CancellationException {
    cancelChecker.checkCanceled();

    List<BuildTargetIdentifier> changedTargets = updateBuildTargets(cancelChecker, token);
    if (!changedTargets.isEmpty()) {
      notifyBuildTargetsChanged(changedTargets, cancelChecker);
    }
  }

  For instance, if we consider the above use case from which getGradleSourceSets method is called and if the cancellation was issued during or after line 76, the cancellation will still be caught in the notifyBuildTargetsChanged method which executes after this and when tokenSource.cancel() is called from the GradleBuildServer class in the catch block, all the operations associated with the token passed will be cancelled.

  It's a best-effort cancellation

Reply:
I doubt that...

Since you are creating new token source for every request/notification. They are all independent. Which means that, cancellation triggered from reload won't affect others because they are different tokens.

Reply:
We need to find a way to persist the relationship between each protocol cancel checker and the gradle cancel token.

For example, you can have a kind of registry holding a map to store the relationship. Meanwhile, a separate thread will keep check the cancel status for each of the cancel checker. If one is cancelled, then call `cancel()` for the token. And you may also pay attention to maintain the map cache, all the out-of-date entries need to be removed.

This is a rough idea for your reference, but there might be solutions better than this...